blog/_posts/2020-08-12-building-root-with-distcc.md

```diff
@@ -0,0 +1,85 @@
+---
+title: "Building ROOT fast - distributed compilation with distcc"
+layout: archive
+author: Stefan Wunsch
+---
+
+Hello there, here is Stefan with a short guide on how you can compile ROOT distributed with distcc!
+
+ROOT is a huge project with more than 24 thousand files and above 4.6 million lines of code (counted with [`cloc`](https://github.com/AlDanial/cloc)). Consequently, a full rebuild from scratch takes a while! To give some numbers:
```

Comment:
I guess this is counting LLVM as part of ROOT? That is a bit misleading w.r.t. the actual size of ROOT.

I get:

```
~/W/R/root (print_jitted_code) fd --extension cxx --extension h --extension C --exclude interpreter | wc -l                                                                   
7812                                                                                                                                                                          
~/W/R/root (print_jitted_code) fd --extension cxx --extension h --extension C  | wc -l                                                                                        
10525                                                                                   
```

Similarly I get 1.7M SLOC with `cloc --force-lang cpp (fd --exclude interpreter --extension cxx --extension h --extension C)` (I might be doing something wrong/different, but 4.6M seems to be an overestimation?).

Reply:
It's with LLVM, yep! Well, I don't know what to write here. At the end we compile almost 5 MLOC, so it's not wrong ;) Even though ROOT itself has probably only 2 MLOC, as you showed. I agree, a little bit misleading, but I'm not sure how to write it differently/better.

Reply:
Even with LLVM I get 10k source files, not 24k.

Something like this maybe:

> ROOT itself consists of about 1.7M lines of code in almost 8k source files. On top of that, our C++ interpreter is based on clang and LLVM, so that is another large project that is compiled as a dependency when building ROOT.

Reply:
```
[stefan@archi root-dev]$ cloc . --exclude-dir llvm
   17090 text files.
   16400 unique files.                                          
    3300 files ignored.

github.com/AlDanial/cloc v 1.86  T=14.71 s (988.7 files/s, 251009.1 lines/s)
---------------------------------------------------------------------------------------
Language                             files          blank        comment           code
---------------------------------------------------------------------------------------
C++                                   5809         299361         363421        1327958
C/C++ Header                          5472         127392         166787         535624
C                                      302          40062          41787         240538
XML                                    800           5435           1622         146246
JavaScript                              77          15350           5517          55943
Python                                 496          13614          16067          49145
Markdown                               222          13981              1          46897
CMake                                  501           4152           5601          25610
HTML                                   197           2329            194          20372
Bourne Shell                           101           1868           1941          17176
Objective C++                           26           3572           3088          13250
Fortran 77                               4            351            540          12570
make                                   284           4847           3798          11416
TeX                                     82           1403           1550          10552
Windows Module Definition                9            181              4           6159
CSS                                     21            566            197           4907
CUDA                                    22           1016           1712           4306
reStructuredText                        33            980            359           3041
m4                                      16            331            506           2355
Objective C                              5            157             89            834
C Shell                                  5             98            136            489
YAML                                     8             68             63            441
Jupyter Notebook                        12              0           2344            370
awk                                      1              1             94            355
JSON                                    10              2              0            288
DOS Batch                                2             35              5            240
Windows Resource File                    4              9              5            173
diff                                     7             31            156            161
Bourne Again Shell                       4             47             28            141
DTD                                      3             25             38            135
SVG                                      1              1              1             80
Fish Shell                               1             10              9             40
NAnt script                              1              1              0             14
SQL                                      1              0              0             14
SWIG                                     2              4              0             14
R                                        1              2              3              8
Perl                                     1              3              4              7
XSLT                                     1              2              7              7
D                                        2              2              0              6
---------------------------------------------------------------------------------------
SUM:                                 14546         537289         617674        2537882
---------------------------------------------------------------------------------------
```

Reply:
i think 14k files and 2.5 MLOC is fair!